subt/tools/validator.py

```diff
@@ -0,0 +1,138 @@
+"""
+  Validate pose3d against ground truth from ignition simulator
+"""
+import datetime
+
+from osgar.logger import LogReader, lookup_stream_id
+from osgar.lib.serialize import deserialize
+import subt.tools.ignstate as ign
+from subt.trace import distance3D
+
+
+MAX_SIMULATION_DURATION = 3700  # 1hour with a small buffer
+
+SIM_TIME_STREAM = 'rosmsg.sim_time_sec'
+ORIGIN_STREAM = 'rosmsg.origin'
+
+
+def evaluate_poses(poses, gt_poses, time_step_sec=1.0):
+    if len(poses) == 0 or len(gt_poses) == 0:
+        return []
+    time_limit = max(poses[0][0], gt_poses[0][0])
+    end_time = min(poses[-1][0], gt_poses[-1][0])
+
+    i, j = 0, 0
+    arr = []
+    while time_limit <= end_time:
+        while poses[i][0] < time_limit:
+            i += 1
+        while gt_poses[j][0] < time_limit:
+            j += 1
+        dist = distance3D(poses[i][1:], gt_poses[j][1:])
+        diff = [a - b for a, b in zip(poses[i][1:], gt_poses[j][1:])]
+        arr.append((time_limit, dist, *diff))
+        time_limit += time_step_sec
+    return arr
+
+
+def ign2arr(ign_poses, robot_name):
+    """Convert Ignition state poses into array"""
+    arr = []
+    for timestamp, data in ign_poses:
+        if robot_name in data:
+            pos = data[robot_name]
+            arr.append((timestamp.total_seconds(), pos.x, pos.y, pos.z))
+    return arr
+
+
+def osgar2arr(poses):
+    arr = []
+    for timestamp, data in poses:
+        arr.append((timestamp.total_seconds(), *data[0]))
+    return arr
+
+
+def read_pose3d(filename, pose3d_name, seconds=MAX_SIMULATION_DURATION):
+    stream_id = lookup_stream_id(filename, pose3d_name)
+    sim_time_id = lookup_stream_id(filename, SIM_TIME_STREAM)
+    poses = []
+    sim_time = 0
+    for dt, channel, data in LogReader(filename, only_stream_id=[stream_id, sim_time_id]):
+        value = deserialize(data)
+        if sim_time > seconds:
+            break
+        if channel == sim_time_id:
+            sim_time = value
+        else:  # pose3d
+            poses.append((datetime.timedelta(seconds=sim_time), value))
```

Comment:
This adds many poses with the same `sim_time`. Is that intended? I thought we want only the first pose after `sim_time` change.

Reply:
It was originally intended because I was not sure if the resolution of `sim_time` changed. It did not and last week we decided that there is no need to increase the logging frequency. So now I would maybe reduce it to single reading after the `sim_time` changed.

Reply:
Can you please make this change? Together with my other suggestion we can then completely ignore the `sim_time` resolution.